spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/LazyWeightedServiceInstanceList.java

```diff
@@ -43,35 +40,30 @@ class LazyWeightedServiceInstanceList extends AbstractList<ServiceInstance> {
 	private final Object expandingLock = new Object();
 
 	LazyWeightedServiceInstanceList(List<ServiceInstance> instances, int[] weights) {
-		this.instances = instances;
-		this.weights = weights;
-		this.init();
+		this.init(instances, weights);
 	}
 
-	private void init() {
-		// Calculate the greatest common divisor (GCD) of weights, max weight and the
+	private void init(List<ServiceInstance> instances, int[] weights) {
+		// Calculate the greatest common divisor (GCD) of weights, and the
 		// total number of elements after expansion.
-		int greatestCommonDivisor = 0;
-		int max = 0;
+		int gcd = 0;
```

Comment:
divide-by-zero is ensured by `WeightedServiceInstanceListSupplier`, that instances is not empty and all weights are greater than 0

Reply:
indeed, but I'd actually like to be able to return 0 in my weight functions.

But not that important for this PR :) 
Feel free to skip over my comments here

Reply:
It's never a good idea that able weight to be 0, how do we deal with it? Ignore it or remove it above all? :)

Reply:
I wouldn't say never :)

Imagine a migration scenario where you gradually divert traffic to new instances but suddenly discover a bug and you want to only send traffic to old instances again. Sure, you could kill the instances, but there might be reasons to not want to do that. Possibly the instances are not even under your control

But my comments are derailing the topic :)
I will address this another time perhaps

Reply:
In general, for all the different load-balancing approaches, we always assume that any instances can be chosen - they might just have a lesser probability of being chosen. For example, if you select by zone preference, we'll get you matching instances for the zone if we can, but if not, we'll get you one selected from the entire set or for the retryable scenario, we will try to retry on a different instance than the failing one, but if we don't get it we'll just use whatever is available. This would be a whole different feature that we might discuss with the and consider whether it should be even implemented separately, so I would not worry about it at this point.